Frameworks/CoreGraphics/CGPath.mm

```diff
@@ -239,45 +202,117 @@ void CGPathAddLineToPoint(CGMutablePathRef path, const CGAffineTransform* m, flo
 
     if (pathObj->_count + 1 >= pathObj->_max) {
         pathObj->_max += 32;
-        pathObj->_components = (pathComponent*)IwRealloc(pathObj->_components, pathObj->_max * sizeof(pathComponent));
+        pathObj->_components = (CGPathElement*)IwRealloc(pathObj->_components, pathObj->_max * sizeof(CGPathElement));
     }
 
-    pathObj->_components[pathObj->_count].type = pathComponentLineTo;
-    pathObj->_components[pathObj->_count].point.x = x;
-    pathObj->_components[pathObj->_count].point.y = y;
+    pathObj->_components[pathObj->_count].type = kCGPathElementAddLineToPoint;
+
+    pathObj->_components[pathObj->_count].points = (CGPoint*)IwCalloc(1, sizeof(CGPoint));
+    pathObj->_components[pathObj->_count].points[0] = { x, y };
 
     pathObj->_count++;
 }
 
+CGFloat _CGPathControlPointOffsetMultiplier(CGFloat angle) {
+    return (4.0f / 3.0f) * tan(angle / 4.0f);
+}
+
 /**
  @Status Caveat
  @Notes transform property not supported
 */
 void CGPathAddArcToPoint(CGMutablePathRef path, const CGAffineTransform* m, float x1, float y1, float x2, float y2, float radius) {
-    if (m) {
-        assert(0);
+    bool isEmpty = CGPathIsEmpty(path);
+
+    if (isEmpty) {
+        return;
     }
 
-    CGPathRef pathObj = path;
+    CGPoint curPathPosition = CGPathGetCurrentPoint(path);
 
-    if (pathObj->_count + 1 >= pathObj->_max) {
-        pathObj->_max += 32;
-        pathObj->_components = (pathComponent*)IwRealloc(pathObj->_components, pathObj->_max * sizeof(pathComponent));
+    double x0, y0;
+    double dx0, dy0, dx2, dy2, xl0, xl2;
+    double san, n0x, n0y, n2x, n2y, t;
+
+    x0 = curPathPosition.x;
+    y0 = curPathPosition.y;
+
+    dx0 = x0 - x1;
```

Comment:
I copied this function directly from CGContextImpl::CGContextAddArcToPoint, which appears to be copied to/from CGContextCairo::CGContextAddArcToPoint. Ideally this logic would live in CGPath and CGContext would call into CGPath so it could be removed from the other locations. I understand cairo can complicate that particular implementation. How much do you want me to handle in this pull request? If I increase readability here and fix any bugs, how would you like to handle this code in the other 2 spots?